### DISCUSSION TITLE
Need Assistance with Firmware Modification for Head Unit: Progress Update and Questions

### DISCUSSION THREAD

**Discussion by al3emsy** [rgerganov/gen5fw#17] [Q&A] [answered]
Hello everyone,

I hope you're doing well. I have been working on modifying the firmware for my head unit, specifically to install the Middle East firmware on my USA-spec car, it is a 2019 Kia Cadenza SXL, with harman/kardan audio system. I have made some progress and would like to share an update on what I have done so far. Additionally, I have a few questions that I hope the community can help me with.

Progress Update:

1. I successfully edited the system.ext4 file to enable ADB and converted it into a system.img file.
2. I obtained the partition.dat file provided in the forum and incorporated it into the firmware directory.
3. I created a folder called "security_force" within the firmware directory and encrypted a total of 7 files inside it: encrypt_boot.img, encrypt_lk.rom, encrypt_partition.dat, and encrypt_recovery.img, encrypt_splash.img, encrypt_system.img, and file_info.

Next Steps:
Now, I am planning to copy the "security_force" directory onto a USB drive formatted with FAT32 as per the instructions provided. I will then proceed with the secret recovery mechanism by holding the POWER and MAP buttons while powering on the head unit or using a toothpick to reset the unit.

Questions:

1. Is the sequence of steps I have described correct? Is there anything I have missed or should be aware of?
2. In case something goes wrong during the update process and I end up in a boot loop, how can I recover the old system? Are there any precautions I should take?

Thank you in advance for your support and guidance. I look forward to your valuable insights and suggestions.

Best regards

![Untitled](https://github.com/rgerganov/gen5fw/assets/134560566/d2035108-f489-4756-bda5-a5ac3f649fdf)

worth mentioning but not sure if it is of any relevance, when checking both MES and USA firmware's on navigation updater, they stand the same format suggesting what [rgerganov](https://github.com/rgerganov/gen5fw/commits?author=rgerganov) said about the units being one but having different firmwares.
the MES update stands at almost 7gb
![Untitled1](https://github.com/rgerganov/gen5fw/assets/134560566/967259ea-f42a-4557-913b-c69acb097f1a)
and the USA version stands at 21gb
![Untitled2](https://github.com/rgerganov/gen5fw/assets/134560566/d433aaef-6e9f-43de-b2e5-e2cdd21bf615)

P.S. now that I am writing this, I am actually wondering, the file security_force contains MES encrypted files, standing a shy 431mb, how is this supposed to turn the headunit to support MES, does forcing this modified  MES files later allow you to go to the official navigation updater and update the original MES firmware?

**Comment by rgerganov** [comment]
> * Is the sequence of steps I have described correct? Is there anything I have missed or should be aware of?

These are the correct steps.

> * In case something goes wrong during the update process and I end up in a boot loop, how can I recover the old system? Are there any precautions I should take?

You can try to create a security_force update with the original USA firmware if the modified one doesn't work. But I cannot give you any guarantees. This method worked for me and failed for others, so use it at your own risk.

> P.S. now that I am writing this, I am actually wondering, the file security_force contains MES encrypted files, standing a shy 431mb, how is this supposed to turn the headunit to support MES, does forcing this modified MES files later allow you to go to the official navigation updater and update the original MES firmware?

The security_force package doesn't contain navigation. Once you flash it, you need to install the original MES package which comes with the navigation.

**Comment by al3emsy** [comment]
Thank you for your response and clarification. I appreciate your confirmation that the sequence of steps I have described is correct.

Regarding the recovery process in case of a boot loop, I understand that creating a security_force update with the original USA firmware is a potential solution. I will keep in mind that this method may not guarantee success and should be used at my own risk.

In my previous post, I mentioned that the security_force package contains MES encrypted files. However, I now understand that this package does not include the navigation feature. To enable navigation, I would need to install the original MES package, which comes with the navigation functionality. Thank you for clarifying this.

I have an additional question: In the event that something goes wrong during the update process, you mentioned creating a new folder called "force_update" with the original USA files. Should these files be left unencrypted? My understanding is that I would extract the contents of the update_package.zip, then extract the update.zip, and finally copy the resulting six files into the "security_force" folder. A missing 7th file would be file_info. Please correct me if I'm mistaken.

Furthermore, if I proceed with the update and use the USB outlet to install the security_force file, should I remove the SD card before proceeding? After the MES firmware is successfully flashed, I plan to install the full MES firmware on the SD card and then reinstall it into the head unit. Does this sound like the correct process?

Lastly, as a precautionary measure, I am currently searching for a replacement MES head unit in case any unexpected issues arise. If the modifications work out as planned, I can return the replacement unit. I will take this as an opportunity to keep you in loop with how it works out with me

P.S. how can I reach out to Ali Al-Rawi.

Thank you for your time and assistance. I greatly appreciate your help in guiding me through this firmware modification process.

Best regards,

**Reply by rgerganov** [reply]
> Should these files be left unencrypted? My understanding is that I would extract the contents of the update_package.zip, then extract the update.zip, and finally copy the resulting six files into the "security_force" folder.

No. The security force update mechanism expect encrypted files.

> Furthermore, if I proceed with the update and use the USB outlet to install the security_force file, should I remove the SD card before proceeding?

I am not sure. I guess it doesn't matter.

> P.S. how can I reach out to Ali Al-Rawi.

Send me an email.

**Comment by al3emsy** [comment]
https://github.com/rgerganov/gen5fw/commits?author=rgerganov
so this is a quick update regarding my progress, I have successfully flashed the MES firmware onto my headunit thanks to your thorough guide and steps, I was presented by the same issue mentioned by other members where menu text is manage applications, you mentioned this was a normal occurrence when changing regions so I went with it, I also updated with the latest original MES firmware and now everything is functioning as normal and all is working great, however I have two arising issues now, speedometer is still in miles, and it does not seem to be affected by the headunit update, I am thinking what if I flash the Canadian firmware, could that have an impact on the speedometer changing it to Km/h? then if that works I can flash back to MES firmware, its just a floating thought and I wanted to know your insights at the matter. Moreover because now I have MES firmware and because my car is equipped with UVO, this service is not available in MESW so I get a little UVO icon with a cross on it indicating it is not working, I also noticed that radio was not working, upon looking around I found out that the UVO module also has the radio functions with it, the system starts UVO then radio, if the system fails to start UVO radio does not work, I am thinking to fix this by removing pins from the UVO motherboard until I can turn off UVO but not the radio, I don't know if this theory is correct or not.

I have been thinking is UVO disabled because it is not in GCC or is there something wrong? 

your insights would be of great value!

**Comment by rgerganov** [comment]
Thanks a lot for confirming the update process worked (at least to some extent) for you.

Where do you see speed as miles/h instead of km/h? If you are talking about the dashboard behind the wheel, I don't think this is part of the head unit.

> I am thinking to fix this by removing pins from the UVO motherboard until I can turn off UVO but not the radio, I don't know if this theory is correct or not.

I am not familiar with UVO but you can certainly try this. You can also decompile `HKMC_RadioGE.odex` and see if there is any option or switch which disables the radio.

Good luck!

**Comment by al3emsy** [comment, marked as answer]
So quick update about my situation, after some thorough digging I came to the conclusion that radio has no sound had something to do with the UVO, after my flash I had a red circle cross with UVO in it indicating UVO was not working, I tried to disable telematics but that only made the UVO sign disappear, but radio still has no sound, so after some digging around I came to understand that for vehicles equipped with UVO/Bluelink, there is a secondary board on top of the main, this board contains cellular signal and Sirius XM, simply disconnecting the power feed did not solve the problem, so the next step was to completely remove the secondary board, and viola! it works, now I have sound in radio!

The way I understand it is that, @rgerganov  method, for me at least was foul proof, it worked like a charm followed the scripts and steps provided by him, so a big special thanks to your efforts, my radio issue was in no relation with his technique, it was a compatibility issue with the MES firmware given that it is unable to use the mobis board. I hope this comes in handy if anyone faces the same issues forward on, and if someone would ever find a way to overcome this obstacle without physically removing the board, that is also welcome.

Once again thank you for you inputs and replies to both @rgerganov and @MurzzZ!

**Reply by rgerganov** [reply]
Thank you for sharing your experience!
